### PR TITLE
Fix make_group_from_vertices for tensor product elements

### PR DESCRIPTION
**Pull request by alexfikl**
The resulting `nodes` were 2d even for a 2d surface embedded in 3d space.

### CODE REVIEW

File: meshmode/mesh/generation.py [hunk 1]
```diff
@@ -372,8 +370,8 @@ def make_group_from_vertices(vertices, vertex_indices, order,
                 vdm[i, j] = np.prod(vertex_ref**func_tuple)
 
         # shape: (dim, nelements, nvertices)
-        coeffs = np.empty((dim, nelements, nvertices))
-        for d in range(dim):
+        coeffs = np.empty((ambient_dim, nelements, nvertices))
```
Comment (alexfikl) [inline meshmode/mesh/generation.py:373]: Basically this is the actual "fix", the rest just moves things around.

File: meshmode/mesh/generation.py [hunk 2]
```diff
@@ -344,21 +345,18 @@ def make_group_from_vertices(vertices, vertex_indices, order,
...
-                break
-            if nvertices < 2**dim:
-                raise ValueError("invalid number of vertices for tensor-product "
-                        "elements, must be power of two")
-            dim += 1
-
-        from modepy.quadrature.jacobi_gauss import legendre_gauss_lobatto_nodes
-        from modepy.nodes import tensor_product_nodes
-        unit_nodes = tensor_product_nodes(dim, legendre_gauss_lobatto_nodes(order))
+        dim = int(np.log2(nvertices))
```
Comment (inducer) [inline meshmode/mesh/generation.py:348]: This seems to round by truncation, which is probably not what you intend:
```
>>> int(np.float64(2.99))
2
```
I sort of even prefer the crappy "iterative log".

File: meshmode/mesh/generation.py [hunk 3]
```diff
@@ -372,8 +370,8 @@ def make_group_from_vertices(vertices, vertex_indices, order,
                 vdm[i, j] = np.prod(vertex_ref**func_tuple)
 
         # shape: (dim, nelements, nvertices)
```
Comment (inducer) [inline meshmode/mesh/generation.py:372]: This comment is now inaccurate.

File: meshmode/mesh/generation.py (hunk 1 above)
Comment (inducer) [reply meshmode/mesh/generation.py:373]: Thx!

File: meshmode/mesh/generation.py [hunk 4]
```diff
@@ -308,12 +308,12 @@ def make_curve_mesh(curve_f, element_boundaries, order,
 # {{{ make_group_from_vertices
 
 def make_group_from_vertices(vertices, vertex_indices, order,
-        group_factory=None):
-    # shape: (dim, nelements, nvertices)
+        group_factory=None, unit_nodes=None):
```
Comment (inducer) [inline meshmode/mesh/generation.py:311]: It's a bit weird that this `group_factory` has nothing to do with the group factories in `poly_element`. Maybe rename as `group_cls` to better document?

**Review by inducer — COMMENTED** [review]
Thanks for working on this! Looks good mostly, just a few wrinkles.

File: meshmode/mesh/generation.py (hunk 2 above)
Comment (alexfikl) [reply meshmode/mesh/generation.py:348]: Silly floating point numbers! How about using [bit_length](https://docs.python.org/3/library/stdtypes.html#int.bit_length)?

File: test/test_meshmode.py [hunk 5]
```diff
@@ -49,6 +49,14 @@
 logger = logging.getLogger(__name__)
 
 
+def acf():
+    import pyopencl as cl
+    from meshmode.array_context import PyOpenCLArrayContext
+    context = cl._csc()
+    queue = cl.CommandQueue(context)
+    return PyOpenCLArrayContext(queue)
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply test/test_meshmode.py:57]: I kind of do, if that's ok with you!

[the remark this answers is not in the document]
Comment (inducer) [reply test/test_meshmode.py:57]: Go for it!

File: meshmode/mesh/generation.py (hunk 2 above)
Comment (inducer) [reply meshmode/mesh/generation.py:348]: Fine by me.

File: meshmode/mesh/generation.py (hunk 4 above)
Comment (alexfikl) [reply meshmode/mesh/generation.py:311]: Yeah, sure. Should I keep `group_factory` with a deprecation warning?

Comment (inducer) [reply meshmode/mesh/generation.py:311]: Yeah, that sounds good.

File: test/test_meshmode.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexfikl) [reply test/test_meshmode.py]: ebb82a9

File: meshmode/mesh/generation.py (hunk 2 above)
Comment (alexfikl) [reply meshmode/mesh/generation.py]: ebb82a9

File: meshmode/mesh/generation.py (hunk 4 above)
Comment (inducer) [reply meshmode/mesh/generation.py]: Yeah, we'll need to get those eventually. Baby steps. #57

**Comment by inducer** [comment]
Thanks!

**Pull request merged**